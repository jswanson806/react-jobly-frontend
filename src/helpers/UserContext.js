// Import the necessary modules from React
import React, { createContext, useState } from 'react';

// Define the initial state for the user
const INITIAL_USER_STATE = {};

// Create a new context for the user
export const UserContext = createContext();

// Create a provider component for the user context
export const UserProvider = ({ children }) => {
  // Set up state to store the user object
  const [user, setUser] = useState(INITIAL_USER_STATE);

  // Render the provider component and pass the user object and setter to the context
  return (
    <UserContext.Provider value={{ user, setUser }}>
      {children}
    </UserContext.Provider>
  );
};

