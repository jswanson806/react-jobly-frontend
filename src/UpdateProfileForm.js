import React, { useContext, useState } from "react";
import { useHistory } from "react-router-dom";
import AppContext from "./helpers/AppContext.js";
import './UpdateProfileForm.css';
import { UserContext } from "./helpers/UserContext.js";

/**
 * UpdateProfileForm Component
 *
 * Renders a form to update user profile information and handles the update functionality.
 *
 * State:
 * - formData: Tracks the state of the form data.
 *
 * Hooks:
 * - useContext: Accesses the updateUserProfile function from the AppContext and user information from the UserContext.
 *
 * Functions:
 * - handleChange: Updates the formData state as the user types in the input fields.
 * - handleSubmit: Submits the form, calls the updateUserProfile function, resets form inputs, and redirects to the homepage.
 */


const UpdateProfileForm = () => {

    const {updateUserProfile} = useContext(AppContext);
    const {user} = useContext(UserContext)
    // initial value of formData
    const INITIAL_STATE = {
        username: user.username,
        firstName: user.firstName,
        lastName: user.lastName,
        email: user.email
    }

    const [formData, setFormData] = useState(INITIAL_STATE);

    const history = useHistory();

    // update the form data as the user types in the input
    const handleChange = (e) => {
        const { name, value } = e.target;
        // update formData state with input field name and value of input field
        setFormData(formData => ({
          ...formData,
          [name]: value,
        }));
      }

    // submit the form data 
    const handleSubmit = async (e) => {
        e.preventDefault();
        // set the input as the value of 'name' in filters
        const newUserInfo = {
            username: formData.username,
            firstName: formData.firstName,
            lastName: formData.lastName,
            email: formData.email
        }
        // call updateUserProfile from parent App.js
        updateUserProfile(newUserInfo);
        // reset form inputs
        setFormData(INITIAL_STATE);
        // redirect to homepage
        history.push('/')
    }

    return (
        <div className="profile-card">
            <h1 className="profile-title">Profile</h1>
            <div>
                <form className="Profile-form" data-testid="Profile-form" onSubmit={handleSubmit}>
                    <label htmlFor="username" className="profile-label">Username</label>
                    <input type="text"
                            id="username"
                            name="username"
                            data-testid="Profile-form-username-input"
                            value={formData.username}
                            disabled
                            tabIndex="-1"
                            className="profile-input"
                    />
                    <label htmlFor="firstName" className="profile-label">First Name</label>
                    <input type="text"
                            id="firstName"
                            name="firstName"
                            data-testid="Profile-form-firstName-input"
                            value={formData.firstName}
                            onChange={handleChange}
                            className="profile-input"
                    />
                    <label htmlFor="lastName" className="profile-label">Last Name</label>
                    <input type="text"
                            id="lastName"
                            name="lastName"
                            data-testid="Profile-form-lastName-input"
                            value={formData.lastName}
                            onChange={handleChange}
                            className="profile-input"
                    />
                    <label htmlFor="email" className="profile-label">Email</label>
                    <input type="text"
                            id="email"
                            name="email"
                            data-testid="Profile-form-email-input"
                            value={formData.email}
                            onChange={handleChange}
                            className="profile-input"
                    />
                    <button type="submit" 
                            data-testid="Profile-form-button"
                            className="profile-button"
                    >
                        Save Changes
                    </button>
                </form>
            </div>
        </div>

    );
};

export default UpdateProfileForm;