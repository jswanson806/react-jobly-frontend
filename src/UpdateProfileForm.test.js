import React from "react";
import {act, render, screen, fireEvent} from '@testing-library/react';
import UpdateProfileForm from './UpdateProfileForm.js';
import AppContext from "./helpers/AppContext.js";
import { Router, MemoryRouter } from "react-router-dom";
import { createMemoryHistory } from 'history';
import '@testing-library/jest-dom';
import { UserContext } from "./helpers/UserContext.js";


const updateUserProfile = jest.fn();
const user = {
    username: "testuser",
    password: "password",
    firstName: "firstName",
    lastName: "lastName",
    email: "test@email.com"
}

describe("login form smoke and snapshot tests", () => {

    it("should render without crashing", async () => {

        await act(async () => {render(
            <UserContext.Provider value={{user}}>
                <AppContext.Provider value={{updateUserProfile}}>
                    <MemoryRouter>
                        <UpdateProfileForm />
                    </MemoryRouter>
                </AppContext.Provider>
            </UserContext.Provider>
        )
        });
    });

    it("should match snapshot", async() => {
        let asFragment;

        await act(async () => {const {asFragment: fragment} = render(
            <UserContext.Provider value={{user}}>
                <AppContext.Provider value={{updateUserProfile}}>
                    <MemoryRouter>
                        <UpdateProfileForm />
                    </MemoryRouter>
                </AppContext.Provider>
            </UserContext.Provider>
        )
        asFragment = fragment;
        });

        expect(asFragment()).toMatchSnapshot();
    })
})

describe("Handles form input correctly", () => {

    it('updates search input correctly', async () => {
        await act(async () => {render(
            <UserContext.Provider value={{user}}>
                <AppContext.Provider value={{updateUserProfile}}>
                    <MemoryRouter>
                        <UpdateProfileForm />
                    </MemoryRouter>
                </AppContext.Provider>
            </UserContext.Provider>
        )
        });

        const userNameInput = screen.getByTestId("Profile-form-username-input");
        const firstNameInput = screen.getByTestId("Profile-form-firstName-input");
        const lastNameInput = screen.getByTestId("Profile-form-lastName-input");
        const emailInput = screen.getByTestId("Profile-form-email-input");
        // this should not change from 'testuser' because input is disabled
        fireEvent.change(userNameInput, {target: {value: "testUserName"}});
        // these should all change
        fireEvent.change(firstNameInput, {target: {value: "testFirstName"}});
        fireEvent.change(lastNameInput, {target: {value: "testLastName"}});
        fireEvent.change(emailInput, {target: {value: "test@email.com"}});

        expect(userNameInput).toHaveValue("testuser");
        expect(firstNameInput).toHaveValue("testFirstName");
        expect(lastNameInput).toHaveValue("testLastName");
        expect(emailInput).toHaveValue("test@email.com");

    })

    it('handles form submission correctly', async () => {
        const history = createMemoryHistory();
        history.push = jest.fn();
        await act(async () => {render(
            <UserContext.Provider value={{user}}>
                <AppContext.Provider value={{updateUserProfile}}>
                    <Router history={history}>
                        <UpdateProfileForm />
                    </Router>
                </AppContext.Provider>
            </UserContext.Provider>
        )
        });

        const form = screen.getByTestId("Profile-form")

        const userNameInput = screen.getByTestId("Profile-form-username-input");
        const firstNameInput = screen.getByTestId("Profile-form-firstName-input");
        const lastNameInput = screen.getByTestId("Profile-form-lastName-input");
        const emailInput = screen.getByTestId("Profile-form-email-input");
        // this should not change from 'testuser' because input is disabled
        fireEvent.change(userNameInput, {target: {value: "testUserName"}});
        // these should all change
        fireEvent.change(firstNameInput, {target: {value: "testFirstName"}});
        fireEvent.change(lastNameInput, {target: {value: "testLastName"}});
        fireEvent.change(emailInput, {target: {value: "test1@email.com"}});

        // submit the form
        fireEvent.submit(form);

        // check for function calls
        expect(updateUserProfile).toHaveBeenCalledTimes(1);
        expect(updateUserProfile).toHaveBeenCalledWith({
            username: "testuser",
            firstName: "testFirstName",
            lastName: "testLastName",
            email: "test1@email.com"
        });
        expect(history.push).toHaveBeenCalledTimes(1);
        expect(history.push).toHaveBeenCalledWith('/');

    })

})