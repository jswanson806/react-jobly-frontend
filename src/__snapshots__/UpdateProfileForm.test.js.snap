// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login form smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="profile-card"
  >
    <h1
      class="profile-title"
    >
      Profile
    </h1>
    <div>
      <form
        class="Profile-form"
        data-testid="Profile-form"
      >
        <label
          class="profile-label"
          for="username"
        >
          Username
        </label>
        <input
          class="profile-input"
          data-testid="Profile-form-username-input"
          disabled=""
          id="username"
          name="username"
          tabindex="-1"
          type="text"
          value="testuser"
        />
        <label
          class="profile-label"
          for="firstName"
        >
          First Name
        </label>
        <input
          class="profile-input"
          data-testid="Profile-form-firstName-input"
          id="firstName"
          name="firstName"
          type="text"
          value="firstName"
        />
        <label
          class="profile-label"
          for="lastName"
        >
          Last Name
        </label>
        <input
          class="profile-input"
          data-testid="Profile-form-lastName-input"
          id="lastName"
          name="lastName"
          type="text"
          value="lastName"
        />
        <label
          class="profile-label"
          for="email"
        >
          Email
        </label>
        <input
          class="profile-input"
          data-testid="Profile-form-email-input"
          id="email"
          name="email"
          type="text"
          value="test@email.com"
        />
        <button
          class="profile-button"
          data-testid="Profile-form-button"
          type="submit"
        >
          Save Changes
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
